pkg/auth/oauth/handlers/interfaces.go

```diff
@@ -25,6 +26,50 @@ type AuthenticationRedirector interface {
 	AuthenticationRedirect(w http.ResponseWriter, req *http.Request) (err error)
 }
 
+// AuthenticationRedirectors is a collection that provides a list of
+// acceptable authentication mechanisms
+type AuthenticationRedirectors struct {
+	Count         int
+	Names         []string
+	RedirectorMap map[string]AuthenticationRedirector
+}
+
+// Add a name and a matching redirection routine to the list
+// The entries added here will be retained in FIFO ordering.
+func (ar *AuthenticationRedirectors) Add(name string, redirector AuthenticationRedirector) {
+	// Initialize the map the first time
+	if ar.RedirectorMap == nil {
+		ar.RedirectorMap = make(map[string]AuthenticationRedirector, 1)
+	}
+
+	// Be overprotective and check whether this name is already in use
+	// This should be impossible, as uniqueness is tested by the config
+	// validator, but it's always best to catch mistakes.
+
+	sanitizedName := name
+	i := 2
+
+	// If this name already exists in the map, add a number to it
+	// We will start at 2 for human comprehension
+	for ar.RedirectorMap[sanitizedName] != nil {
+		sanitizedName = name + strconv.Itoa(i)
+		i++
+	}
+
+	ar.Names = append(ar.Names, sanitizedName)
+	ar.RedirectorMap[sanitizedName] = redirector
+	ar.Count++
+}
+
+// Get the AuthenticationRedirector associated with a name.
+// Also returns a boolean indicating whether the name matched
+func (ar *AuthenticationRedirectors) Get(name string) (redirector AuthenticationRedirector, exists bool) {
+	var val AuthenticationRedirector
```

Comment:
That didn't work. At least on Go 1.6, I get an error that I'm trying to only return a single value when the function expects to return two. Doing this explicitly resolved it.